query-service-impl/src/main/java/org/hypertrace/core/query/service/ExecutionContext.java

```diff
@@ -209,6 +184,70 @@ private void extractColumns(List<String> columns, Expression expression) {
     }
   }
 
+  private long getPeriodInSeconds(String[] timeSeriesPeriod) {
+    String period = timeSeriesPeriod[0];
+    // add cases for other units when added
+    switch (timeSeriesPeriod[1]) {
+      case "MILLISECONDS":
+        return Long.parseLong(period) / 1000;
+      default:
+        return Long.parseLong(period);
+    }
+  }
+
+  private void treeTraversal(
+      Filter filter,
+      AtomicLong lessThanFilter,
+      AtomicLong greaterThanFilter,
+      String timeFilterColumn) {
+
+    if (lessThanFilter.longValue() >= 0 && greaterThanFilter.longValue() >= 0) {
```

Comment:
So rather than passing some mutable longs down the traversal tree, this would read much cleaner IMO if you avoided recursing into leaf nodes (this started as pseudocode and got a little more formal, but may still need some work):

```java

private Optional<Duration> findTimeRangeDuration(Filter filter, String timeFilterColumn) {
  if (!filter.operation.equals(AND)) {
     return Optional.empty(); // a range must be AND'd together, so we can abort any OR/leaf nodes
  }

  Optional<Long> timeRangeStart =  filter.getChildFilterList()
    .filter(childFilter -> this.isMatchingFilter(childFilter, timeFilterColumn, List.of(GT, GTE)))
    .map(matchingFilter -> matchingFilter.getRhs().getLiteral().getValue().getLong())
    .findFirst();


  Optional<Long> timeRangeEnd =  filter.getChildFilterList()
    .filter(childFilter -> this.isMatchingFilter(childFilter, timeFilterColumn, List.of(LT, LTE)))
    .map(matchingFilter -> matchingFilter.getRhs().getLiteral().getValue().getLong())
    .findFirst();

  if (timeRangeStart.isPresent() && timeRangeEnd.isPresent()) {
   // Is the column unit part of the pinot config? This is an unsafe assumption
   return Duration.ofMillis(timeRangeEnd.get() - timeRangeStart.get());
  }

  return filter.getChildFilterList()
    .map(filter -> this.findTimeRangeDuration(filter, timeFilterColumn))
    .flatMap(Optional::stream)
    .findFirst();
}

private boolean isMatchingFilter(Filter filter, String column, Collection<Operator> operators) {
    return Objects.equals(column, filter.getLhs().getColumnIdentifier().getColumnName()) &&
      operators.anyMatch(operator -> Objects.equals(operator, filter.getOperator());
}